### PR TITLE
Allow viewers to join org

### PR DESCRIPTION
**Pull request by amartinj**
Fixes [#1373](https://github.com/CartoDB/support/issues/1373)

### CODE REVIEW

File: app/controllers/signup_controller.rb [hunk 1]
```diff
@@ -183,25 +183,20 @@ def check_organization_quotas
...
-  def valid_email_invitation_token
+  def invitation
+    return @invitation if @invitation
     email = (params[:user] && params[:user][:email]) || params[:email]
     token = params[:invitation_token]
-    if email && token
-      invitations = Carto::Invitation.query_with_valid_email(email).where(organization_id: @organization.id).all
-      @invitation ||= invitations.find { |i| i.token(email) == token }
-    end
+    return if !email || !token
```
Comment (javitonino) [inline app/controllers/signup_controller.rb]: `unless email && token` ?

File: app/controllers/signup_controller.rb [hunk 2]
```diff
@@ -183,25 +183,20 @@ def check_organization_quotas
 
   def load_mandatory_organization
     load_organization
-    render_404 and return false unless @organization && (@organization.signup_page_enabled || valid_email_invitation_token?)
+    render_404 and return false unless @organization && (@organization.signup_page_enabled || invitation)
```
[the remark this answers is not in the document]
Comment (javitonino) [reply app/controllers/signup_controller.rb]: You can do `return render_404 unless ...` and Hound will calm down.

File: spec/requests/signup_controller_spec.rb [hunk 3]
```diff
@@ -132,6 +132,18 @@
       response.status.should == 200
       response.body.should include("quota_in_bytes not enough disk quota")
     end
+
+    it 'signs user up as viewer even if all non-viewer seats are taken' do
+      @fake_organization = FactoryGirl.create(:organization_with_users, whitelisted_email_domains: [])
+      @fake_organization.seats = @fake_organization.users.count(&:builder?)
+      @fake_organization.viewer_seats = 10
+      @fake_organization.save
```
Comment (javitonino) [inline spec/requests/signup_controller_spec.rb]: I'm starting a petition to always use `save!` in tests (otherwise, your fixture set up may fail and you don't even notice)

File: spec/requests/signup_controller_spec.rb [hunk 4]
```diff
@@ -132,6 +132,18 @@
...
+      owner = Carto::User.find(@fake_organization.owner.id)
+      invitation = Carto::Invitation.create_new(owner, ['wadus@wad.us'], 'Welcome!', true)
+      Organization.stubs(:where).returns([@fake_organization])
```
Comment (javitonino) [inline spec/requests/signup_controller_spec.rb]: Why is this needed? It seems too aggressive. I think you can replace this with `host! "#{@fake_organization.name}.localhost.lan"` which will configure the signup URL to be in the proper domain for this to work.

File: spec/requests/signup_controller_spec.rb [hunk 5]
```diff
@@ -132,6 +132,18 @@
       response.status.should == 200
       response.body.should include("quota_in_bytes not enough disk quota")
     end
+
+    it 'signs user up as viewer even if all non-viewer seats are taken' do
+      @fake_organization = FactoryGirl.create(:organization_with_users, whitelisted_email_domains: [])
+      @fake_organization.seats = @fake_organization.users.count(&:builder?)
```
Comment (javitonino) [inline spec/requests/signup_controller_spec.rb]: There is `@fake_organization.builder_users.count` but this should be the same.

File: spec/requests/signup_controller_spec.rb (hunk 4 above)
Comment (amartinj) [reply spec/requests/signup_controller_spec.rb]: Did not think about this line too much. Used one of the tests above as reference

File: spec/requests/signup_controller_spec.rb (hunk 3 above)
Comment (amartinj) [reply spec/requests/signup_controller_spec.rb]: Yes, I tried to, but it's a sequel model:
```
     Failure/Error: @fake_organization.save!
     NoMethodError:
       undefined method `save!' for #<Organization:0x0000000c96ff08>
```

**Comment by ramiroaznar** [comment]
LGTM!

org
